AP_motors: add dualcopter/Bicopter

IamPete1 opened the pull request:
This is more of a rearranging than adding a new frame type. The tail-sitter output mixers for thrust vectoring are moved up to AP_MotorTailsitter. This then allows copter to use them for bicopter. Then rename tailsitter to dual and enable in copter 

The longer term goal is to improve the mixers to work well for a bicopter to benefit thrust vectored tail-sitters. 

I have tested a thrust vectored tail-sitter in plane and a bi copter in copter, both in SITL. There should be no change to tail-sitter behavior.  New bicopter for copter fly's OK but is not fantastically stable but did successfully complete a auto-tune. 

https://youtu.be/PtJzx_vx7N0

I have not looked that deeply into Copter to see if there is anything else I should change to support the new frame type, possibly I have missed some stuff

rmackay9 commented on ArduCopter/Parameters.cpp:
```diff
@@ -869,7 +869,7 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
     // @DisplayName: Frame Class
     // @Description: Controls major frame class for multicopter component
-    // @Values: 0:Undefined, 1:Quad, 2:Hexa, 3:Octa, 4:OctaQuad, 5:Y6, 6:Heli, 7:Tri, 8:SingleCopter, 9:CoaxCopter, 11:Heli_Dual, 12:DodecaHexa, 13:HeliQuad
+    // @Values: 0:Undefined, 1:Quad, 2:Hexa, 3:Octa, 4:OctaQuad, 5:Y6, 6:Heli, 7:Tri, 8:SingleCopter, 9:CoaxCopter, 10:DualCopter, 11:Heli_Dual, 12:DodecaHexa, 13:HeliQuad
```
I think we should call the frame BiCopter instead of DualCopter because it's the more common name used for this frame type (I think) and reduces confusion with Heli_Dual.

rmackay9 commented on libraries/AP_Motors/AP_Motors_Class.h:
```diff
@@ -53,7 +53,9 @@ class AP_Motors {
         MOTOR_FRAME_TYPE_Y6B = 10,
-        MOTOR_FRAME_TYPE_Y6F = 11 // for FireFlyY6
+        MOTOR_FRAME_TYPE_Y6F = 11, // for FireFlyY6
+        MOTOR_FRAME_TYPE_TAILSITTER = 12, // none vectored tailsiters
```
I think maybe we are no longer using MOTOR_FRAME_TYPE_TAILSITTER so we don't need this element 12?

rmackay9 commented on libraries/AP_Motors/AP_MotorsTailsitter.cpp:
```diff
@@ -85,28 +84,29 @@ void AP_MotorsTailsitter::output_to_motors()
+    SRV_Channels::set_output_scaled(SRV_Channel::k_tiltMotorRight, _tilt_right*SERVO_OUTPUT_RANGE);
+
+    // plane outputs for Qmodes are setup here, and written to the HAL by the plane servos loop
+    SRV_Channels::set_output_scaled(SRV_Channel::k_aileron,  -_yaw_in*SERVO_OUTPUT_RANGE);
```
It's existing code so it's OK for this PR but it's a bit weird that the motors library is setting the output ranges for k_aileron, k_elevator, k_rudder and k_throttle.  I would much prefer if one part of the code was responsible for the setup and output of each servo/esc output rather than having this responsibility done in two very separate parts of the code.

IamPete1 replied:
Yeah, i think this is just a peculiarity of how the tail-sitter code works as the same output functions are used by both the copter and plane code. Depending on if its in a forward flight mode these outputs are overwritten.

IamPete1 commented:
I have made the changes. Although I am not sure that I have added the new methods correctly. The bit mask and test expects the code to be using the motor1, motor2... output functions. The tailsitter code uses throttle left, throttle right, tilt left and tilt right. I'm not sure how to convert these into the correct bit masks. 

Also I can't get the renaming of AP_MotorsTailsitter.cpp to AP_MotorsDual.cpp to work. It works fine for AP_MotorsTailsitter.h to AP_MotorsDual.h and keeps all the history as it should. I have left it as AP_MotorsTailsitter.cpp for the time-being but this causes the code to not compile. I used git mv, it worked fine the first time for both files. Any ideas?

IamPete1 commented:
I have got this flying on a real aircraft, it flys! I will do a vid once I have tuned it up abit. 

As it is the motor test doesn't work. I think not using the standard motor1... motor2 ect is causing a few issues.

IamPete1 commented:
I have added the output function to AP_Motors to use the tailsitter output functions. This enables motor test to work correctly and also I hope esc calibration but I haven't tested. I have updated it to use the RC_write functions within the AP_motors library.

This flys on a SITL bicopter but I have yet to check it still works with tailsitters.  I hope to add some  code to reduce average throttle to maintain control. Currently the outputs are simply constrained.

IamPete1 commented:
I have rebased this and i think its ready to go for review, I have tested it on a bicopter and thrust vectored tail-sitter in SITL. May also be worth testing it on a none-vectored tail-sitter I guess, although I don't foresee and issue there.

IamPete1 commented:
I have done some more SITL testing, it works fine for none vectored tail sitters. The motor test works fine in Copter but not in Plane. I cant work out where the error is it seems to enter the motor test sequence because the outputs do not respond correctly when armed, but they do not throttle up as per the usual test.  As far as I can tell if it works on copter it should work on plane so not sure what is going on.

tridge commented:
@IamPete1 nice work! can you do a PR against SITL_Models for the bicopter?
https://github.com/ArduPilot/SITL_Models

tridge commented:
does this work well for people updating from existing parameters on TVBS and tailsitter aircraft?

tridge commented:
one missing piece when running as a copter is the Q_TAILSIT_VHGAIN. That is effectively set to 1 I think. That means you can't autotune in copter and transfer the gains to plane vectored tailsitter.
I think we need a way to set this gain for copter, so you can fly tailsitters as copters and tune them there

IamPete1 commented:
@tridge I have added the bicopter model to the SITL_Models. This should be no change for tail-sitters, my SITL one (I have added this to the SIT_Models too) fly's as it should and I have tested on the CAT-tail-sitter also. 

As you say Q_TAILSIT_VHGAIN is missing, this is the same as using 1. If we wanted to autotune in copter I think we would also have to move the control surface scaling stuff up from tailsitter,cpp. This would require moving Q_TALSIT_VHGAIN and Q_TAILSIT_THSCMX, and maybe others up to the motors library. If we did this then none vectored tail-sitters should also fly under copter.

I did have a look a few mouths ago at make autotune available to plane as Qautotune. I think this would be quite a big job but doable. If you think its a good idea and feasible I could have a go. 

Thanks for the review

IamPete1 commented:
@tridge  @rmackay9  Now Plane has its own auto tune I think this is probably OK to go in as it is?

rmackay9 commented:
Thanks for this.  It's looking pretty good but I think I've found a few issues: 

- no need to add the extra AP_MOTORS_MAX_NUM_ACTUATORS, we can just stick with AP_MOTORS_MAX_NUM_MOTORS.  This definition is just to define the maximum number of motors a frame can have.  A BiCopter only has two motors so it falls well within the limit.
- the APMotorsTailsitter::get_motor_mask() should return a bitmask corresponding to the motor outputs that are being used.  This bitmask is used in the vehicle code to ensure that nothing else (like the gimbal) tries to output on the same PWM output.  This function by the way is far less important than it use to be now that we've got the SERVOx_FUNCTION parameters that mostly stops users from assigning multiple features to use the same output.
- OK to remove the "virtual" in front of the "output_test_seq()" definition.

rmackay9 commented:
I'm quite keen to get BiCopter support in master (my main sponsor has a copter they want to fly with this) so I've gone ahead and made the changes I was thinking of above.

Here's a link to my branch built on top of this PR: https://github.com/rmackay9/rmackay9-ardupilot/commits/iampete1-bicopter

@IamPete1, maybe you could have a look at this and see if it seems OK?  I haven't tried to fly it.

IamPete1 commented:
@rmackay9 The tilt outputs don't seem to be getting sent in your branch, I think I had a similar issue when I was first making this although I cant remember how I fixed it exactly, I have re based this to compare.

IamPete1 commented:
@rmackay9 I think how I have done set_update_rate(), output_test_seq() and :get_motor_mask()  are inline with how it is done for coax and tri. 

I think we need the additional AP_MOTORS_MAX_NUM_ACTUATORS param to use the rc_write(), so if we go back to set_ouput_scaled() this can be removed. I think we will loose the ability to do esc calibration if we go back to set_ouput_scaled()? This would defiantly be a nice feature for all the tailsitter people. 

Great to hear your company wants to fly something on this, I have just about got my bicopter auto-tuning although not completed all the way through yet, I only get 8 or 10 mins flight time. I did make a start on some [updated maths for the controller ](https://github.com/IamPete1/ardupilot/commit/20a2a622068cac648c5f872b5e1c4396c3ab3ddf) (untested) but I wanted to get it tuned as it is to set a bench mark. Weather had been too bad for flying for ages unfortunately so I have not done much on it recently.

rmackay9 commented:
@IamPete1, thanks for the feedback I'll look into the esc calibration and the tilt issue.

I think single-copter is probably not a good guide unfortunately.  Single copter hasn't really been properly updated since we moved to the SRV_Channel library which allows users to configure which outputs are used for a given feature.  Single and CoaxCopter are still using hard-coded outputs.

IamPete1 commented:
@rmackay9 I it looks to me that this, coax and tri are done the same as AP_MotorsMatrix. I think I came up against some issue when I was attempting to do it using set_ouput_scaled() that wouldn't allow motor test to work, but looking at the code now I cant think what It could have been. I also had the tilts not updating issue but again I don't remember how I fixed it. Then I removed all that stuff to go to the rc_write method.

rmackay9 commented:
OK, I found the typo that was probably causing the tilt servos to not move properly.  I also likely got the esc calibration working although I can't say the solution is perfect.  It might be better to instead have AP_MotorsTailsitter.cpp override the AP_MotorsMulticopter::set_throttle_passthrough_for_esc_calibration function and just output to the left and right throttles.

Sorry if I missed it but is there a simulator file somewhere that I could test with?

IamPete1 commented:
great! There is a [real-flight ](https://github.com/ArduPilot/SITL_Models/tree/master/RealFlight/iampete)model for bicopter and tail-sitter that should both work with this. I can test them both on your branch now if you like. Not sure how to test esc calibration tho.

rmackay9 commented:
@IamPete1, if you could test it that would be great! (I'm about to go to sleep).  I can test the ESC calibration using an external servo tester tomorrow.

IamPete1 commented:
@rmackay9 Both this any your branch fly equally well in plane and copter, both do individual motor test in copter. This does motor test sequence wheres as your branch does not. Neither motors tests work on plane, I think this is a unrelated issue with plane, I looked into it before and couldn't find anything obvious. 

I think the sequence motor test issue is not a big deal (it may have been this that made me change to the rc write method) so I will pull your changes into this PR.

IamPete1 commented:
I have pulled your changes to this branch, so this should now be the same as yours. Also I have pulled just the formatting fixes to [this backup branch](https://github.com/IamPete1/ardupilot/tree/bicopter-backup). So the only difference should be that this branch uses set_ouput_scaled() and my 'backup' branch still uses rc_write. I think the only thing left to check is esc calibration for copter and ideally for plane too.  

Thanks

rmackay9 commented:
@IamPete1, great. I've tested both the ESC calibration and also the motor test and they seem to work ok.  

By the way, this is the default motor ordering that this library uses:

Servo Out 1: right throttle
Servo Out 1: right throttle
Servo Out 1: right throttle

rmackay9 commented:
Tridge has also tested it on a plane tailsitter and reported it works so I've gone ahead and merged this.  We have some ongoing questions we can address is follow up PRs if necessary.

IamPete1 commented on ArduPlane/tailsitter.cpp:
```diff
@@ -217,9 +218,11 @@ void QuadPlane::tailsitter_speed_scaling(void)
         SRV_Channel::Aux_servo_function_t::k_aileron,
-        SRV_Channel::Aux_servo_function_t::k_elevator};
+        SRV_Channel::Aux_servo_function_t::k_elevator,
+        SRV_Channel::Aux_servo_function_t::k_tiltMotorLeft,
```
Previous to this the speed scaling is applied to elevator and aileron then the scaled values are mixed to get tilt left and right. Now the tilt left and right is calculated in AP_motors so copter can use it too, so we have to apply the tail sitter scaling after mixing it. This gives the same result ( (x + y) * z = x * z + y *z ) so should be no change in tail sitter behaviour.

BkMak16 commented:
Hi

I just want to find out what arguments I use if I wanted to start the bicopter using sitl. 

I think it would be 

sim_vehicle.py -v ArduCopter -f ????

Or would it be 

sim_vehicle.py -v ArduPlane -f quad-plane 
param set FRAME_CLASS 1
param set TILT_TYPE 3

Preferred to start the bicopter as copter rather than plane. Is is possible? 
Sorry if there's any mistakes I'm on my phone.

IamPete1 commented:
@BkMak16 There is no bicopter physics model for the build in SITL simulator. You can use realflight with this model https://github.com/ArduPilot/SITL_Models/tree/master/RealFlight/iampete/Bicopter Just launch a copter simulation and load in the params.

BkMak16 commented:
Okay thank you. 

You've saved me alot of attempted coding. 
I then have another question, is it possible to have a custom simulated multi rotor, say modified parameters of a quadrotor

IamPete1 commented:
@BkMak16 not in the built in sim, but you can do whatever physics you like in realflight.

BkMak16 commented:
Thanks. I'll go read up about realflight.

IamPete1 commented:
@BkMak16 https://ardupilot.org/dev/docs/sitl-with-realflight.html

BkMak16 commented:
Awesome. So over the passed day or so I've read up on realflight. I don't think it will fit with what I want so I would rather take the quadcopter into airsim without the bicopter. 

Thanks for all your help!
